### PR TITLE
Cartesian Eigenvectors

### PR DESCRIPTION
**Pull request by rebeccafair**
Addresses #86

### CODE REVIEW

**Comment by rebeccafair** [comment]
Ok @g5t @mducle  I think this is finally about ready for review. I have tested the wheels with Euphonic using the new interface syntax and it all seems to work!

I ended up keeping the `RotatesLike` name rather than changing it to `ValueType`, as I feel `ValueType` is a bit generic, and `RotatesLike` at least makes it clear it is used in rotating values to the 1st BZ?

I wasn't sure what kind of errors etc. you want for the `switch case` so I kept it as simple as possible for now. Let me know if it needs changing, once we've decided on what should happen I can write some tests.

Do I need to change the docs too?

**Comment by g5t** [comment]
Thanks for the effort to get this pulled together. I don't see any problems with the implementation and the tests/builds succeed, so I think everything looks good!

Regarding the documentation: you've updated doc strings in the source which get pulled into the [Sphinx pages](https://brille.github.io/pull/repository/index.html) (I need to fix the action to put that documentation someplace more useful, but that's not a problem for this PR). Do you think the documentation needs anything else for a user to understand the use of the two enumerations?

I think the change you made to avoid re-generating the numpy archive for `test_5_gamma` is fine. A small comment explaining why the field `nacl['grid_vectors_elements']` is not read might be nice, but a tool like `git-blame` is probably sufficient to remind ourselves later if we forget. 
I'd argue against updating the archived data anyway since a) the binary file is stored in the brille repository, and I don't think we should duplicate it for such a minor change and b) my fetching method doesn't verify the file against, e.g., a hash, so old versions of the project might fail this test due to the API change without warning if we replace the file.

Adding a test for the `RotatesLike::Gamma`, `LengthUnit::angstrom` case might be useful, but it's similar enough to `RotatesLike::Gamma`, `LengthUnit::real_lattice` to not be critical.

**Review by g5t — APPROVED** [review]
I think this PR is fine as-is, but you're free to add more documentation, comments, or tests if you think they're necessary and/or easy to implement.

**Comment by rebeccafair** [comment]
@g5t Ah I didn't realise the pbdoc changes showed up so well in the Sphinx pages! I've had a look and can't see elsewhere anything that contradicts the new interface so I think the docs are fine. I've added some tests for the new Cartesian and error cases. Note that you can add an incompatible combination of `RotatesLike` and `LenghUnit` via `fill`, and it won't error until `ir_interpolate_at` is called, is this ok?

If you're happy with the changes I'm happy for this to be merged.

File: src/interpolator.hpp [hunk 1]
```diff
@@ -385,12 +391,39 @@ class Interpolator{
...
+                brille::utils::mul_mat_mat(tdbl0.data(), 3u, rt.lattice().to_xyz(LengthUnit::angstrom).data(), ps.data(j));
+                // rot_cart = tdbl0*inv(lattice) = lattice*rot*inv(lattice)
+                brille::utils::mul_mat_mat(rot_cart[j].data(), 3u, tdbl0.data(), rt.lattice().from_xyz(LengthUnit::angstrom).data());
+              }
+              return this->rip_gamma(x,q,rt,rot_cart,r,invr,nth);
+            }
+            default: throw std::runtime_error("LengthUnit, RotatesLike combination not implemented");
+        }
+      }
+      default: throw std::runtime_error("LengthUnit, RotatesLike combination not implemented");
```
Comment (mducle) [inline src/interpolator.hpp:426]: I think the above is all fine, but you could also use bit-masks and do it like:

```cpp
#define ROTATESLIKE_VECTOR (1 << 0)
#define ROTATESLIKE_PSEUDO (1 << 1)
#define ROTATESLIKE_GAMMA (1 << 2)
#define LENGTH_ANGSTROM (1 << 16)
#define LENGTH_INVANGS (1 << 32)
#define LENGTH_REAL (1 << 64)
#define LENGTH_RECIP (1 << 128)

int bit_mask = (rotlike_ == RotatesLike::vector ? ROTATESLIKE_VECTOR : 0) | 
               (rotlike_ == RotatesLike::pseudovector ? ROTATESLIKE_PSEUDO : 0) |
               (rotlike_ == RotatesLike::gamma ? ROTATESLIKE_GAMMA : 0) |
               (lenunit_ == LengthUnit::real_lattice ? LENGTH_REAL  : 0) |
               (lenunit_ == LengthUnit::reciprocal_lattice ? LENGTH_RECIP  : 0) |
               (lenunit_ == LengthUnit::angstrom ? LENGTH_ANGSTROM  : 0) |
               (lenunit_ == LengthUnit::inverse_angstrom ? LENGTH_INVANG  : 0);

switch (bit_mask) {
    case (LENGTH_REAL & ROTATESLIKE_VECTOR):       return this->rip_real(x,ps,r,invr,nth);
    case (LENGTH_REAL & ROTATESLIKE_PSEUDO):       return this->rip_axial(x,ps,r,invr,nth);
    case (LENGTH_REAL & ROTATESLIKE_GAMMA):        return this->rip_gamma(x,q,rt,ps.getall(),r,invr,nth);
    case (LENGTH_RECIP & ROTATESLIKE_VECTOR):      return this->rip_recip(x,ps,r,invr,nth);
    case (LENGTH_ANGSTROM & ROTATESLIKE_VECTOR): {
        // Convert rotation matrix from fractional to Cartesian
        std::vector<std::array<double,9>> rot_cart(ps.size());
        std::array<double,9> tdbl0;
        for (size_t j=0; j<ps.size(); j++){
          // tdbl0 = lattice*rot
          brille::utils::mul_mat_mat(tdbl0.data(), 3u, rt.lattice().to_xyz(LengthUnit::angstrom).data(), ps.data(j));
          // rot_cart = tdbl0*inv(lattice) = lattice*rot*inv(lattice)
          brille::utils::mul_mat_mat(rot_cart[j].data(), 3u, tdbl0.data(), rt.lattice().from_xyz(LengthUnit::angstrom).data());
        }
        return this->rip_gamma(x,q,rt,rot_cart,r,invr,nth);
    }
    default: throw std::runtime_error("LengthUnit, RotatesLike combination not implemented");
}
```

It's more C than C++-like though... but might make it easier to add more cases later?

**Review by mducle — APPROVED** [review]

File: src/interpolator.hpp (hunk 1 above)
Comment (g5t) [reply src/interpolator.hpp:426]: I like the cleaner `switch` statement, but I'm not sure that I like macros. 
A third (further `namespace`ed) `enum class RotatesLike_LengthUnit` and an appropriate definition of `brille::interpolation::RotatesLike_LengthUnit operator&(brille::RotatesLike, brille::LengthUnit)`  would enable the same clean logic without losing type information, but might be overkill:

```cpp
switch (rotlike_ & lenunit_) {
  case (brille::RotatesLike::vector & brille::LengthUnit::real_lattice):       return  rip_real(x, ps, r, invr, nth);
  case (brille::RotatesLike::pseudovector & brille::LengthUnit::real_lattice): return rip_axial(x, ps, r, invr, nth);
  case (brille::RotatesLike::Gamma & brille::LengthUnit::real_lattice):        return rip_gamma(x, q, rt, ps.getall(), r, invr, nth);
  case (brille::RotatesLike::vector & brille::LengthUnit::reciprocal_lattice)  return rip_recip(x, ps, r, invr, nth);
  case (brille::RotatesLike::vecotr & brille::LengthUnit::angstrom): {
    ...
  }
  default throw std::runtime_error("LengthUnit, RotatesLike combination not implemented");
}
```

I'll think about this some more, and might implement one variant in the near future; but for now I think this possible change should not prevent merging this PR as it achieves the overall goal of implementing #86

**Pull request merged**